### PR TITLE
Keep test pr on top redo

### LINKED ISSUE #1016: Merge queue: after running tests, keep a PR at the top of the queue.
It often happens in the merge queue:

PR A is being tested.
PR B is added into the queue.
PR B is higher priority than PR A.
Head flakes.
PR A tests finish.
Head goes green.
PR B starts getting tested.

Bug report: at that last step, due to my optimization in #1006, if PR A were checked again, it'd get merged immediately. Instead, PR B is checked, then PR A tests are run AGAIN.

Desired behavior is to leave a PR that's just been tested at the top of the queue until head goes green. Then: if it's mergable, merge it. Otherwise, sort the queue again.

### PR DESCRIPTION
**Pull request by apelisse**
Try to fix #1016

### CODE REVIEW

File: mungegithub/mungers/submit-queue.go [hunk 1]
```diff
@@ -927,110 +933,138 @@ func (sq *SubmitQueue) handleGithubE2EAndMerge() {
...
-		sq.githubE2ERunning = nil
-		delete(sq.githubE2EQueue, keys[0])
-		sq.Unlock()
+func (sq *SubmitQueue) mergePullRequest(obj *github.MungeObject) {
+	obj.MergePR("submit-queue")
+	sq.updateMergeRate()
+	sq.SetMergeStatus(obj, merged)
+}
+
+func (sq *SubmitQueue) reportFailedPullRequest(obj *github.MungeObject, err string) {
```
Comment (eparis) [inline mungegithub/mungers/submit-queue.go]: what's the rational for this function?  "eventually" ?

Comment (apelisse) [inline mungegithub/mungers/submit-queue.go]: No it's useless. I was focused on success/failure symmetry. I didn't realize this was useless ... Will remove!

File: mungegithub/mungers/submit-queue.go [hunk 2]
```diff
@@ -927,110 +933,138 @@ func (sq *SubmitQueue) handleGithubE2EAndMerge() {
...
+		return nil
+	} else if baseSHA, gotBaseSHA := obj.GetSHAFromRef(baseRef); !gotBaseSHA {
+		return nil
+	} else {
+		return &submitQueueInterruptedObject{obj, headSHA, baseSHA}
 	}
 }
 
-func (sq *SubmitQueue) doGithubE2EAndMerge(obj *github.MungeObject) {
+func (sq *SubmitQueue) doGithubE2EAndMerge(obj *github.MungeObject) bool {
```
Comment (eparis) [inline mungegithub/mungers/submit-queue.go]: Can you put a comment to explain the meaning of the return code?

Comment (apelisse) [inline mungegithub/mungers/submit-queue.go]: Sure! I'll resubmit quickly

File: mungegithub/mungers/submit-queue.go (hunk 1 above)
Comment (lavalamp) [inline mungegithub/mungers/submit-queue.go]: +1

File: mungegithub/mungers/submit-queue.go (hunk 2 above)
Comment (lavalamp) [inline mungegithub/mungers/submit-queue.go]: +1

**Comment by lavalamp** [comment]
@apelisse if you can fix comments, squash, & push, maybe we can get this in today. Thanks!

We probably need to test this behavior-- maybe you can add a todo.

**Comment by lavalamp** [comment]
LGTM - I kicked travis. I think my flakesync test may be too intense for travis's overcomitted environment?

**Comment by lavalamp** [comment]
@eparis @mikedanese let's merge this next time we push. I probably won't do a push until I'm confident in #1061.

**Pull request merged**